docs/NNNN-schema-registry.md

```diff
@@ -0,0 +1,152 @@
+
+* Where an OCFL is being generated as a snapshot for export or backup purposes, it is reasonable to build or check the completeness of the schema registry as each OCFL object is created. Thus the exported root will contain all necessary schemata.
+
+* Where an OCFL provides the persistence layer for a live repository, it is recommended that a background process discovers (or is informed) of new OCFL object/versions. This process can asynchronously process the object contents, retrieve the required schemata and perform necessary registrations with the schema registry.
+
+Since schemata are typically identified by the URL of the specific version referenced, a hashed representation of the identifier is used to create a safe filename for storage. MD5 is likely to be sufficient although other algorithms may be configured. The extension's inentory stores a map of the hashed identifiers, thus allowing easy access to the appropriate schema from unmodified OCFL objects.
+
+### Registering a schema with the extension
+
+* New OCFL objects/versions are inspected for JSON, XML files. 
+* Within these files, references to external schema ($schema / XML-DTD etc) are extracted
+* The local filename is derived using the specified schemaDigestAlgorithm.
+* schemata_inventory.json is checked for this key. If not already present in the schema repository:
+  * A copy of the remote schema is retrieved, and stored in the schemata/ directory with this name.
+  * The schemata_inventory.json is updated:
+    * a ‘manifest’ object entry is created for fixity.
+    * an entry in the ‘schemaMap’ object is created to provide a mapping of the hashed identifiers to their original value.
+  * The schema_inventory.json’s sidecar file must also be updated.
+
+
+## Example
+
+Given OCFL objects:
+
+item1
+: with descriptive metadata in content/item1.xml 
+
+```
+<!DOCTYPE rdf:RDF SYSTEM "http://dublincore.org/specifications/dublin-core/dcmes-xml/2001-04-11/dcmes-xml-dtd.dtd">
+...
+```
+item2
+: with descriptive metadata in content/item2.json
+
+```
+{
+  "$schema" : "http://schemata.hasdai.org/historic-persons/historic-person-entry-v1.0.0.json"
+...
+```
+
+The extension is configured via config.json as shown:
+
+```
+{
+  "extensionName" : "NNNN-schema-registry",
+  "schemaDigestAlgorithm": "md5"
+}
+```
+
+The identifiers' md5 digests are calculated -
+
+The schemata are then registered in the extension's schema_inventory.json as shown below. 
+
+
+```
+{
+  "digestAlgorithm": "sha512",
+  "manifest": {
+    "91da2b...9c8": [
+      "schemata/40cdd53d9a263e5466b8954d82d23daa"
+    ],
+    "31f53b...7ff": [
+      "schemata/95d751340dcdc784fd759dbc7ddb9633"
+    ]
+  }
+  "schemaMap": {
+    "40cdd53d9a263e5466b8954d82d23daa": "http://dublincore.org/specifications/dublin-core/dcmes-xml/2001-04-11/dcmes-xml-dtd.dtd",
+    "95d751340dcdc784fd759dbc7ddb9633": "http://schemata.hasdai.org/historic-persons/historic-person-entry-v1.0.0.json"
+  }
```

Comment:
A few questions here:

1. Are these keys case sensitive?
2. It seems like `schemaMap` does not map URLs to paths, but rather URLs to digests which are understood to be file names in the schemata directory. There is a notable difference here with what's done in the `manifest` map. Personally, it would seem more useful to me if the `schemaMap` used the URL as the key and had a path to the file, eg `schemata/40cdd53d9a263e5466b8954d82d23daa`, as the value. This both makes it easier to retrieve the files as well as makes it possible to relate a URL to its file digest in the `manifest`.

Reply:
While it does mean a litte extra hashing, I'm a little averse to using the identifier (essentially any arbitrary string) as a object key. 

Interested to see what others think here. The question of case sensitivty depends on what we decide to do, but I'd lean towards 'yes to be on the safe-side'.

Reply:
The schemaMap is not intented to override common sense. Of course you could simply hash the key and check for the presence of the file. However, as I commented elsewhere, verifying the identifier matches is prudent in case of collision (unlikely, but possibly a mallicious attack vector...)

Having thought further, I conclude that trying to make the inventory OCFL-like is unnecessary (and a potential source of confusion). As you suggest there is no need to accomoate the full features of an OCFL inventory file - ie version de-duplication, logical paths etc are not required - the 'content' of the registry is simply a flat collection of schemata. 

So I'm going to suggest simply 1 "manifest" structure - indexed by identifier hash, with sub-keys of both the identifier and the sha512 (ie fixity) digest, ie:

```
"manifest" : {
  "40cdd53d9a263e5466b8954d82d23daa" : {
    "identifier" : "http://dublincore.org/specifications/dublin-core/dcmes-xml/2001-04-11/dcmes-xml-dtd.dtd",
    "digest" : "91da2b...9c8"
  },
  ... 
}
```

This approach allows easy checking the integrity of the stored schemata and provides an index of all schemata in use. Storing the original identifier is important for several reasons: 

1 - need to verify the hash of the identifier doesn't collide (as mentioned above) before depositiing a new one
2 - it provides a useful reference point for users/processes inspecting the OCFL
3 - in a DR situation, a lost schema can be identified and possibly retrieved

Would the structure above (replacing manifest and schemaMap) work for you? If so will make a PR to reflect this structure.

Reply:
Personally, I like this proposal more. It would be great to have some other opinions other than just mine though.

Reply:
@pwinckles I agree that the new proposal is more elegant and neater to parse too.